feature: register shelter

eTaphee opened the pull request:
<!-- 필요한 경우 팀원들과 의논하여 양식을 변경할 수 있습니다. -->
<!-- PR은 상세히 적어주시는게 좋습니다. -->
<!-- 이해를 돕기위한 이미지를 써도 좋습니다. -->

## Changes
<!-- 어떤점이 변경되었는지 적어주세요. -->
보호소 등록 기능을 구현했습니다.
- 보호소를 등록한 사용자가 호출하게 되면 예외가 발생합니다.
- 등록한 사용자 기준은 SHLETER role 여부에 따라 결정됩니다.

최대한 저희가 정한 규칙을 따르게 작성했습니다.
- Record 대신 Class
- 유효성 검사 시 최대한 커스텀 어노테이션을 지양하고 기본 제공 어노테이션 활용
- 테스트 작성 시 any 지양

## Background
<!-- 이 PR이 진행된 배경을 적어주세요. -->
보호소 등록 API가 필요

## Discuss
<!-- 토론할 내용이 있다면 적어주세요. -->
어제의 멘토링 시간 이후로 단위 테스트에 대한 근거없는 확신(?)이 생긴거 같아요. 😂
테스트 작성하는게 재밌는거 같아요..

테스트 네이밍 같은 경우 멘토님이 말해주신 양식을 벗어나지 않는 한도로 개인이 작성하면 되지 않을까 싶습니다.

## Issues
<!-- 관련된 이슈를 #{issueNumber}를 통해 태그해주세요. -->
<!-- ex) 해결한 이슈: #1, #2 -->
<!-- ex) 구현이 필요한 이슈: #3, #4 -->
#57 

## Execute
<!-- 실행된 결과에 대해 적어주세요. -->
<!-- 어떻게 테스트를 진행했는지, 어떤 검토를 거쳤는지 적어주시면 좋습니다. -->
<!-- 테스트 코드는 중요합니다! -->
<img width="1125" alt="image" src="https://github.com/banzzokee/banzzokee-back/assets/43163409/d13c3cac-eae4-4dc6-a2f2-e08dfaedc121">


## Reference
<!-- 참조한 레퍼런스가 있다면 적어주세요. -->

eTaphee committed:
feature: 보호소 등록 구현

eTaphee committed:
test: 보호소 등록 테스트 작성

eTaphee committed:
feature: 보호소 승인 구현

eTaphee committed:
test: 테스트 컨테이너 비활성

eTaphee committed:
test: 테스트 코드 수정

Shelter에 UserId 관계를 추가하여 영향받는 테스트 코드 수정

eTaphee committed:
test: 보호소 승인 테스트 작성

jnjeaaaat commented on src/main/java/homes/banzzokee/domain/shelter/exception/UserAlreadyRegisterShelterException.java:
```diff
@@ -0,0 +1,19 @@
+@Getter
+public class UserAlreadyRegisterShelterException extends CustomException {
+
+  private final Long userId;
```
userId, shelterId 가 따로 예외처리를 할때 쓰이는 곳이 있는건가요??

eTaphee replied:
따로 쓰이는 곳은 아직까지 없습니다.

추가한 이유는 나중에 로그분석 할 때 필요하지 않을까 해서 추가한거긴 한데, 당장 필요없으면 지우는게 맞을거 같네요.

jnjeaaaat commented on src/main/java/homes/banzzokee/domain/shelter/service/ShelterService.java:
```diff
@@ -0,0 +1,59 @@
+        .build());
+  }
+
+  private void throwIfUserAlreadyRegisterShelter(User user) {
```
1. 보호소 등록 신청을 했는데 승인이 안난 경우
2. 승인까지 되어서 등록이 되어있는 경우

두가지 경우를 다르게 처리하면 좋을것 같은데 어떠신가요???

eTaphee replied:
그런 예외 케이스도 있었네요 반영하겠습니다.
고맙습니다.

jnjeaaaat commented on src/main/java/homes/banzzokee/domain/shelter/service/ShelterService.java:
```diff
@@ -0,0 +1,59 @@
+
+    S3Object uploadedImage = uploadShelterImage(shelterImg);
+
+    user.registerShelter(Shelter.builder()
```
shelterRepository.save() 가 없는것 같습니다.. 다른곳에서 작성하신 걸까요??
제가 못찾은 거라면 죄송합니다..

eTaphee replied:
매핑 시  @OneToOne(fetch = LAZY, cascade = PERSIST)
cascade 옵션을 PERSIST로 설정했기 때문에, user의 shelter가 없는 경우 자동으로 생성합니다.

jnjeaaaat replied:
오 그렇군요..! 답변 감사합니다~!

eTaphee committed:
refactor: PR 피드백 반영

- 불필요한 Exception 파라미터 제거
- 보호소 등록 시 승인되지 않은 보호소 예외 케이스 추가

eTaphee committed:
Merge branch 'feature/register-shelter' into feature/approve-shelter

eTaphee committed:
feature: 보호소 수정 구현

eTaphee committed:
test: 보호소 수정 테스트 작성

eTaphee committed:
feature: 보호소 삭제 구현

- 삭제된 보호소에 대해 인증이 되지 않도록 수정
- 보호소를 삭제 후 다시 등록될 수 있도록 구현
- 보호소 삭제 로직
  - 보호소의 deletedAt, verified 업데이트
  - 사용자의 SHELTER 권한 삭제
  - 보호소가 참여한 채팅방에서 shelter_id = null 업데이트

eTaphee committed:
test: 보호소 삭제 테스트 작성

eTaphee committed:
refactor: 보호소 서비스 테스트 리팩터링

eTaphee committed:
test: 보호소 승인 테스트 케이스 추가

nadoran78 commented on src/main/java/homes/banzzokee/domain/shelter/controller/ShelterController.java:
```diff
@@ -0,0 +1,28 @@
+
+  private final ShelterService shelterService;
+
+  @PostMapping
```
@PostMapping(value = "", consumes = {MediaType.APPLICATION_JSON_VALUE,
      MediaType.MULTIPART_FORM_DATA_VALUE})
이렇게 별도 설정하지 않아도 별 문제가 없나보군요~

nadoran78 commented on src/main/java/homes/banzzokee/domain/shelter/dto/ShelterRegisterRequest.java:
```diff
@@ -0,0 +1,55 @@
+  /**
+   * 연락처
+   */
+  @Pattern(regexp = TEL)
```
pattern을 별도 관리하는 걸 새로 배우네요

nadoran78 commented on src/main/java/homes/banzzokee/global/validator/constant/RegexPattern.java:
```diff
@@ -1,16 +1,10 @@
 
-import lombok.Getter;
-import lombok.RequiredArgsConstructor;
+public final class RegexPattern {
```
enum 타입에서 class로 변경하신 이유가 어떤 것일까요?

eTaphee replied:
커스텀 어노테이션을 사용하면 문제가 되지 않지만..

@Pattern 을 사용하게 되면 regexp 값을 상수로 입력해야 되더라구요.

그래서 enum(String value)는 인스턴스 값이라서 @Pattern에 사용할 수 없어서 변경하게 됐습니다.

enum으로도 @Pattern에 활용할 수 있는 방법을 알고 계실까요?

eTaphee committed:
Merge pull request #64 from banzzokee/feature/delete-shelter

feature: delete shelter

eTaphee committed:
Merge pull request #63 from banzzokee/feature/update-shelter

feature: update shelter

eTaphee committed:
Merge pull request #60 from banzzokee/feature/approve-shelter

feature: verify shelter
